src/sema/expression.rs

```diff
@@ -4260,11 +4260,13 @@ fn assign_single(
                 }
             }
 
+            let ty = ty.deref_any();
+
             Ok(Expression::Assign(
                 *loc,
                 ty.clone(),
                 Box::new(var.clone()),
-                Box::new(val.cast(&right.loc(), ty.deref_any(), true, ns, diagnostics)?),
+                Box::new(val.cast(&right.loc(), ty, true, ns, diagnostics)?),
```

Comment:
Out of curiosity, why is this changed?

Reply:
The problem is that the assignment expression `a = c` should have the same type as the value being assigned to `a`. I first changed it to:

```
Ok(Expression::Assign(
                *loc,
                ty.deref_any().clone(),
                Box::new(var.clone()),
                Box::new(val.cast(&right.loc(), ty.deref_any(), true, ns, diagnostics)?),
                Box::new(val.cast(&right.loc(), ty, true, ns, diagnostics)?),
```
But this means `deref_any()` is being called twice, so avoided that with a let binding.

Reply:
In other words, this changes the second field of Expression::Assign to the right value.

Reply:
Ah yes I get it now :facepalm: thanks